examples/dataflow-csvio/src/main/java/com/google/example/csvio/SortContextualHeadersAndRows.java

```diff
@@ -0,0 +1,427 @@
+
+  /**
+   * A {@link PTransform} responsible for bundling the raw {@link ContextualTextIO#read()} {@link
+   * Row} {@link PCollection} output into the corresponding CSV headers and rows in the setting
+   * where a regular expression header matching is expected.
+   */
+  static class SortMatchedHeaderAndRows extends PTransform<PCollection<Row>, Result> {
+
+    private static final String TAG_BASE =
+        SortContextualHeadersAndRows.TAG_BASE
+            + "/"
+            + SortMatchedHeaderAndRows.class.getSimpleName();
+    private final CSVIOReadConfiguration configuration;
+
+    SortMatchedHeaderAndRows(CSVIOReadConfiguration configuration) {
+      this.configuration = configuration;
+    }
+
+    @Override
+    public Result expand(PCollection<Row> input) {
+      PCollectionView<String> matchHeaderRegexView =
+          input
+              .getPipeline()
+              .apply(
+                  TAG_BASE + "/MatchHeaderRegexView",
+                  Create.of(configuration.getHeaderMatchRegex()))
+              .apply(View.asSingleton());
+
+      PCollection<KV<String, Long>> resourceIdHeaderPosition =
+          input.apply(
+              TAG_BASE + "/" + FindMatchedHeaderFn.class.getSimpleName(),
+              ParDo.of(new FindMatchedHeaderFn(matchHeaderRegexView))
+                  .withSideInput("", matchHeaderRegexView));
+      PCollectionView<Map<String, Long>> mappedHeaderPositionView =
+          resourceIdHeaderPosition.apply(View.asMap());
+      PCollectionTuple pct =
+          input.apply(
+              TAG_BASE + "/" + SortMappedFixedPositionHeadersAndRowsFn.class.getSimpleName(),
+              ParDo.of(new SortMappedFixedPositionHeadersAndRowsFn(mappedHeaderPositionView))
+                  .withSideInput("", mappedHeaderPositionView)
+                  .withOutputTags(HEADERS, TupleTagList.of(ROWS)));
+
+      return new Result(pct);
+    }
+  }
+
+  /**
+   * A {@link PTransform} responsible for bundling the raw {@link ContextualTextIO#read()} {@link
+   * Row} {@link PCollection} output into the corresponding CSV headers and rows in the setting
+   * where the header is expected at the first non-empty line of the CSV file.
+   */
+  static class SortMinimumHeaderAndRows extends PTransform<PCollection<Row>, Result> {
+
+    private static final String TAG_BASE =
+        SortContextualHeadersAndRows.TAG_BASE
+            + "/"
+            + SortMinimumHeaderAndRows.class.getSimpleName();
+
+    @Override
+    public Result expand(PCollection<Row> input) {
+      PCollection<KV<String, Long>> resourceIdRecordOffsetKV =
+          input
+              .apply(
+                  TAG_BASE + "/" + IsNotEmptyFn.class.getSimpleName(),
+                  Filter.by(new IsNotEmptyFn()))
+              .apply(
+                  TAG_BASE + "/MapElements/" + RowKVFn.class.getSimpleName(),
+                  MapElements.via(new RowKVFn()));
+
+      PCollection<KV<String, Long>> minimumResourceIdRecordOffsetKV =
```

Comment:
Need reevaluation of calculating recordnum during the cases when [withRecordNumMetadata()](https://beam.apache.org/releases/javadoc/2.38.0/org/apache/beam/sdk/io/contextualtextio/ContextualTextIO.Read.html#withRecordNumMetadata--) is not used and can be set as 1 (i.e default position)